### PR TITLE
Make Allocator.h less silly (no creepy "proto" object).

### PR DESCRIPTION
**Pull request by ywkaras**
Replaces https://github.com/apache/trafficserver/pull/6153 .

### CODE REVIEW

File: src/traffic_server/InkAPI.cc [hunk 1]
```diff
@@ -1454,6 +1454,7 @@ HttpHookState::Scope::operator++()
 void
 HttpHookState::Scope::clear()
 {
+  _hooks = nullptr;
```
Comment (ywkaras) [inline src/traffic_server/InkAPI.cc]: This change is needed with or without the changes to the class allocators.  I think the only reasons we are not seeing crashes in master currently is that, by luck, the prototype object has zeros in this field.

**Comment by ywkaras** [comment]
Here's some sample code to compare construction with placement new versus copying from a prototype object:  https://godbolt.org/z/PPsY_C .

Note that for construction, when it's faster, the compiler will generate and use its own (partial) prototype object (LCP4_0).

File: src/traffic_server/InkAPI.cc (hunk 1 above)
Comment (ywkaras) [reply src/traffic_server/InkAPI.cc]: @duke8253

Comment (ywkaras) [reply src/traffic_server/InkAPI.cc]: The change to this file is now a separate PR:  https://github.com/apache/trafficserver/pull/6275

**Comment by ywkaras** [comment]
[approve ci FreeBSD]

**Comment by ywkaras** [comment]
[approve ci docs]

File: include/tscore/Allocator.h [hunk 2]
```diff
@@ -142,6 +140,9 @@ template <class C> class ClassAllocator : public Allocator
   void
   free(C *ptr)
   {
+    if (Destruct_on_free) {
```
Comment (masaori335) [inline include/tscore/Allocator.h:143]: Seems a nice idea. I always struggle with how to destroy/free objects when I use `ClassAllocator`.

File: include/tscore/Allocator.h [hunk 3]
```diff
@@ -152,7 +153,7 @@ template <class C> class ClassAllocator : public Allocator
     @param tail pointer to be freed.
     @param num_item of blocks to be freed.
    */
-  void
+  std::enable_if_t<!Destruct_on_free, void>
```
Comment (masaori335) [inline include/tscore/Allocator.h:156]: Is this a guard to do not set `Destruct_on_free` true?

**Comment by masaori335** [comment]
I'm buying this idea. Especially, we can get rid of a hack of `force_VFPT_to_top`.  
I'd like to get more eyeballs, because this is a fundamental change of Allocator.

File: include/tscore/Allocator.h (hunk 3 above)
Comment (ywkaras) [reply include/tscore/Allocator.h:156]: Yes this is a Standard library template that uses "substitution failure is not a error".  if Destruct_of_free is false, the member function will not be defined for the instantiation.

Comment (masaori335) [reply include/tscore/Allocator.h:156]: I'm not a bit familiar with SFINAE. Let me clarify, which is your intention here when `Destruct_on_free == true` ?
a). do not define this function 
or 
b). make a compile error

Background: I got a `'enable_if' cannot be used to disabl
e this declaration` error when I tried setting the flag `true` like `ClassAllocator<UnixNetVConnection, true>`. If your intention is to do not declare/define this function, this error seems not expected.
```
/usr/local/opt/llvm/include/c++/v1/type_traits:542:78: error: no type named 'type' in 'std::__1::enable_if<false, void>'; 'enable_if' cannot be used to disabl
e this declaration
template <bool _Bp, class _Tp = void> using enable_if_t = typename enable_if<_Bp, _Tp>::type;
                                                                             ^~~
../../include/tscore/Allocator.h:156:8: note: in instantiation of template type alias 'enable_if_t' requested here
  std::enable_if_t<!Destruct_on_free, void>
       ^                                                                                                                                                      UnixNetProcessor.cc:334:10: note: in instantiation of template class 'ClassAllocator<UnixNetVConnection, true>' requested here
    vc = THREAD_ALLOC_INIT(netVCAllocator, t);
         ^
/Users/masaori/src/github.com/apache/trafficserver-review/iocore/eventsystem/I_ProxyAllocator.h:114:35: note: expanded from macro 'THREAD_ALLOC_INIT'
#define THREAD_ALLOC_INIT(_a, _t) thread_alloc_init(::_a, _t->_a)
```
( clang version 11.0.0 )

Comment (ywkaras) [reply include/tscore/Allocator.h:156]: Sorry, I did this wrong, should have tested it better.  I'll fix it.

Comment (ywkaras) [reply include/tscore/Allocator.h]: OK, this is much improved now.  The conditionally enabled functions were not necessary and have been eliminated.

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by bryancall** [comment]
I can benchmark this and see if there is any differences in performance.

**Comment by ywkaras** [comment]
Thanks.

**Comment by shinrich** [comment]
I applied this patch to our current production ATS9 build, and it core's on start up with the following stack.  Appears that something is not getting correctly initialized on reading in the cache.  The member variable od of type OpenDirEntry appears to be bogus value.  I'm setting up an ASAN build to hopefully get more details.

```
(gdb) bt
#0  0x00000000006cdaae in remove (this=0xf8f4bb5c, e=0x2b6f1d20d800) at ../../../../../../_vcs/trafficserver9/iocore/cache/CacheDir.cc:138
#1  OpenDir::close_write (this=0x2b6cea1061e0, cont=cont@entry=0x2b6f1d20d800) at ../../../../../../_vcs/trafficserver9/iocore/cache/CacheDir.cc:138
#2  0x00000000006ede4c in close_write (cont=0x2b6f1d20d800, this=<optimized out>) at ../../../../../../_vcs/trafficserver9/iocore/cache/P_CacheInternal.h:765
#3  CacheVC::openWriteCloseDir(int, Event*) () at ../../../../../../_vcs/trafficserver9/iocore/cache/CacheWrite.cc:1112
#4  0x00000000006f1e86 in Cache::open_write(Continuation*, ats::CryptoHash const*, HTTPInfo*, long, ats::CryptoHash const*, CacheFragType, char const*, int) ()
    at ../../../../../../_vcs/trafficserver9/iocore/cache/CacheWrite.cc:1831
#5  0x00000000006bc925 in CacheProcessor::open_write (this=<optimized out>, cont=cont@entry=0x2b6ceed085c8, expected_size=expected_size@entry=0, key=key@entry=0x2b6ce440d610, 
    request=request@entry=0x2b6ceed06fb0, old_info=<optimized out>, pin_in_cache=0, type=CACHE_FRAG_TYPE_HTTP) at ../../../../../../_vcs/trafficserver9/iocore/cache/Cache.cc:3254
#6  0x00000000005dd2a8 in HttpCacheSM::open_write (this=this@entry=0x2b6ceed085c8, key=key@entry=0x2b6ce440d610, url=url@entry=0x2b6ceed06938, request=request@entry=0x2b6ceed06fb0, 
    old_info=old_info@entry=0x2b6f1d206558, pin_in_cache=<optimized out>, retry=true, allow_multiple=false) at ../../../../../../_vcs/trafficserver9/proxy/http/HttpCacheSM.cc:362
#7  0x0000000000564a23 in HttpSM::do_cache_prepare_action(HttpCacheSM*, HTTPInfo*, bool, bool) () at ../../../../../../_vcs/trafficserver9/proxy/http/HttpSM.cc:4722
#8  0x0000000000582765 in do_cache_prepare_write (this=0x2b6ceed06800) at ../../../../../../_vcs/trafficserver9/proxy/http/HttpSM.cc:4652
#9  HttpSM::set_next_state() () at ../../../../../../_vcs/trafficserver9/proxy/http/HttpSM.cc:7596
#10 0x00000000005756b1 in HttpSM::state_api_callout(int, void*) () at ../../../../../../_vcs/trafficserver9/proxy/http/HttpSM.cc:1547
#11 0x0000000000579c07 in HttpSM::state_api_callback(int, void*) () at ../../../../../../_vcs/trafficserver9/proxy/http/HttpSM.cc:1351
#12 0x000000000051361f in TSHttpTxnReenable () at ../../../../../_vcs/trafficserver9/src/traffic_server/InkAPI.cc:6164
#13 0x00002b6cef802a5d in (anonymous namespace)::cb (contp=<optimized out>, event=<optimized out>, edata=0x2b6ceed06800) at _vcs/quick_filter-9/quick_filter/quick_filter.cc:637
#14 0x00000000004fe1d0 in INKContInternal::handle_event(int, void*) () at ../../../../../_vcs/trafficserver9/src/traffic_server/InkAPI.cc:1096
#15 0x0000000000510a9b in handleEvent (data=0x2b6ceed06800, event=60003, this=0x2b6cd6650a00)
    at /home/shinrich/build-walt/_build/build_release_posix-x86_64_gcc_8/trafficserver9/build/../../../../_vcs/trafficserver9/iocore/eventsystem/I_Continuation.h:167
#16 handleEvent (data=0x2b6ceed06800, event=60003, this=0x2b6cd6650a00)
    at /home/shinrich/build-walt/_build/build_release_posix-x86_64_gcc_8/trafficserver9/build/../../../../_vcs/trafficserver9/iocore/eventsystem/I_Continuation.h:163
#17 APIHook::invoke(int, void*) const () at ../../../../../_vcs/trafficserver9/src/traffic_server/InkAPI.cc:1333
#18 0x00000000005754ac in HttpSM::state_api_callout(int, void*) () at ../../../../../../_vcs/trafficserver9/proxy/http/HttpSM.cc:1475
#19 0x0000000000582abb in HttpSM::set_next_state() () at ../../../../../../_vcs/trafficserver9/proxy/http/HttpSM.cc:7355
#20 0x000000000056ab42 in HttpSM::do_hostdb_lookup() () at ../../../../../../_vcs/trafficserver9/proxy/http/HttpSM.cc:4175
#21 0x0000000000582b5a in HttpSM::set_next_state() () at ../../../../../../_vcs/trafficserver9/proxy/http/HttpSM.cc:7671
#22 0x00000000005756b1 in HttpSM::state_api_callout(int, void*) () at ../../../../../../_vcs/trafficserver9/proxy/http/HttpSM.cc:1547
#23 0x0000000000579c07 in HttpSM::state_api_callback(int, void*) () at ../../../../../../_vcs/trafficserver9/proxy/http/HttpSM.cc:1351
#24 0x000000000051361f in TSHttpTxnReenable () at ../../../../../_vcs/trafficserver9/src/traffic_server/InkAPI.cc:6164
#25 0x00002b6d07cda97e in main_handler (cont=<optimized out>, event=<optimized out>, edata=0x2b6ceed06800) at ../../../../../_vcs/trafficserver9/plugins/regex_revalidate/regex_revalidate.c:453
#26 0x00000000004fe1d0 in INKContInternal::handle_event(int, void*) () at ../../../../../_vcs/trafficserver9/src/traffic_server/InkAPI.cc:1096
#27 0x0000000000510a9b in handleEvent (data=0x2b6ceed06800, event=60015, this=0x2b6cda99b980)
    at /home/shinrich/build-walt/_build/build_release_posix-x86_64_gcc_8/trafficserver9/build/../../../../_vcs/trafficserver9/iocore/eventsystem/I_Continuation.h:167
#28 handleEvent (data=0x2b6ceed06800, event=60015, this=0x2b6cda99b980)
    at /home/shinrich/build-walt/_build/build_release_posix-x86_64_gcc_8/trafficserver9/build/../../../../_vcs/trafficserver9/iocore/eventsystem/I_Continuation.h:163
#29 APIHook::invoke(int, void*) const () at ../../../../../_vcs/trafficserver9/src/traffic_server/InkAPI.cc:1333
#30 0x00000000005754ac in HttpSM::state_api_callout(int, void*) () at ../../../../../../_vcs/trafficserver9/proxy/http/HttpSM.cc:1475
#31 0x0000000000579c07 in HttpSM::state_api_callback(int, void*) () at ../../../../../../_vcs/trafficserver9/proxy/http/HttpSM.cc:1351
#32 0x000000000051361f in TSHttpTxnReenable () at ../../../../../_vcs/trafficserver9/src/traffic_server/InkAPI.cc:6164
#33 0x00002b6cfab731cd in carpLookup(tsapi_cont*, TSEvent, void*) () at _vcs/carp-9/carp/carp.cc:767
#34 0x00000000004fe1d0 in INKContInternal::handle_event(int, void*) () at ../../../../../_vcs/trafficserver9/src/traffic_server/InkAPI.cc:1096
#35 0x0000000000510a9b in handleEvent (data=0x2b6ceed06800, event=60015, this=0x2b6cd6652120)
    at /home/shinrich/build-walt/_build/build_release_posix-x86_64_gcc_8/trafficserver9/build/../../../../_vcs/trafficserver9/iocore/eventsystem/I_Continuation.h:167
#36 handleEvent (data=0x2b6ceed06800, event=60015, this=0x2b6cd6652120)
---Type <return> to continue, or q <return> to quit---
    at /home/shinrich/build-walt/_build/build_release_posix-x86_64_gcc_8/trafficserver9/build/../../../../_vcs/trafficserver9/iocore/eventsystem/I_Continuation.h:163
#37 APIHook::invoke(int, void*) const () at ../../../../../_vcs/trafficserver9/src/traffic_server/InkAPI.cc:1333
#38 0x00000000005754ac in HttpSM::state_api_callout(int, void*) () at ../../../../../../_vcs/trafficserver9/proxy/http/HttpSM.cc:1475
#39 0x0000000000582abb in HttpSM::set_next_state() () at ../../../../../../_vcs/trafficserver9/proxy/http/HttpSM.cc:7355
#40 0x00000000005756b1 in HttpSM::state_api_callout(int, void*) () at ../../../../../../_vcs/trafficserver9/proxy/http/HttpSM.cc:1547
#41 0x0000000000582abb in HttpSM::set_next_state() () at ../../../../../../_vcs/trafficserver9/proxy/http/HttpSM.cc:7355
#42 0x0000000000577bce in HttpSM::state_cache_open_read(int, void*) () at ../../../../../../_vcs/trafficserver9/proxy/http/HttpSM.cc:2553
#43 0x0000000000578a7b in HttpSM::main_handler(int, void*) () at ../../../../../../_vcs/trafficserver9/proxy/http/HttpSM.cc:2615
#44 0x00000000005dd031 in handleEvent (data=0x2b6f1d206400, event=1102, this=0x2b6ceed06800)
    at /home/shinrich/build-walt/_build/build_release_posix-x86_64_gcc_8/trafficserver9/build/../../../../_vcs/trafficserver9/iocore/eventsystem/I_Continuation.h:167
#45 handleEvent (data=0x2b6f1d206400, event=1102, this=0x2b6ceed06800)
    at /home/shinrich/build-walt/_build/build_release_posix-x86_64_gcc_8/trafficserver9/build/../../../../_vcs/trafficserver9/iocore/eventsystem/I_Continuation.h:163
#46 HttpCacheSM::state_cache_open_read(int, void*) () at ../../../../../../_vcs/trafficserver9/proxy/http/HttpCacheSM.cc:114
#47 0x00000000006e7717 in handleEvent (data=0x2b6f1d206400, event=1102, this=0x2b6ceed085c8)
    at /home/shinrich/build-walt/_build/build_release_posix-x86_64_gcc_8/trafficserver9/build/../../../../_vcs/trafficserver9/iocore/eventsystem/I_Continuation.h:167
#48 handleEvent (data=0x2b6f1d206400, event=1102, this=0x2b6ceed085c8)
    at /home/shinrich/build-walt/_build/build_release_posix-x86_64_gcc_8/trafficserver9/build/../../../../_vcs/trafficserver9/iocore/eventsystem/I_Continuation.h:163
#49 CacheVC::callcont (this=this@entry=0x2b6f1d206400, event=event@entry=1102) at ../../../../../../_vcs/trafficserver9/iocore/cache/P_CacheInternal.h:639
#50 0x00000000006e2f8e in CacheVC::openReadStartEarliest(int, Event*) () at ../../../../../../_vcs/trafficserver9/iocore/cache/CacheRead.cc:996
#51 0x00000000006c41f4 in CacheVC::handleReadDone(int, Event*) () at ../../../../../../_vcs/trafficserver9/iocore/cache/Cache.cc:2301
#52 0x00000000006c6b1e in AIOCallbackInternal::io_complete (this=0x2b6f1d206588, event=<optimized out>, data=<optimized out>)
    at /home/shinrich/build-walt/_build/build_release_posix-x86_64_gcc_8/trafficserver9/build/../../../../_vcs/trafficserver9/iocore/aio/P_AIO.h:121
#53 0x00000000007cf2fb in handleEvent (data=0x2b6d25601360, event=1, this=0x2b6f1d206588) at ../../../../../../_vcs/trafficserver9/iocore/eventsystem/I_Continuation.h:167
#54 handleEvent (data=0x2b6d25601360, event=1, this=0x2b6f1d206588) at ../../../../../../_vcs/trafficserver9/iocore/eventsystem/I_Continuation.h:163
#55 EThread::process_event(Event*, int) () at ../../../../../../_vcs/trafficserver9/iocore/eventsystem/UnixEThread.cc:132
#56 0x00000000007cf9ce in EThread::process_queue (this=this@entry=0x2b6cda63a740, NegativeQueue=NegativeQueue@entry=0x2b6ce440e600, ev_count=ev_count@entry=0x2b6ce440e5fc, 
    nq_count=nq_count@entry=0x2b6ce440e5f8) at ../../../../../../_vcs/trafficserver9/iocore/eventsystem/UnixEThread.cc:167
#57 0x00000000007cfe34 in EThread::execute_regular (this=this@entry=0x2b6cda63a740) at ../../../../../../_vcs/trafficserver9/iocore/eventsystem/UnixEThread.cc:227
#58 0x00000000007d04c6 in execute (this=0x2b6cda63a740) at ../../../../../../_vcs/trafficserver9/iocore/eventsystem/UnixEThread.cc:332
#59 EThread::execute (this=0x2b6cda63a740) at ../../../../../../_vcs/trafficserver9/iocore/eventsystem/UnixEThread.cc:310
#60 0x00000000007ce8a9 in spawn_thread_internal (a=0x2b6cd2adf840) at ../../../../../../_vcs/trafficserver9/iocore/eventsystem/Thread.cc:92
#61 0x00002b6cd0df8ea5 in start_thread (arg=0x2b6ce4410700) at pthread_create.c:307
#62 0x00002b6cd1b2e8dd in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
(gdb) 
```

**Comment by shinrich** [comment]
The asan build crashed too.  One stack the same as the earlier two.  The other stack different but also due to uninitialized ok member variable.

```
(gdb) bt
#0  0x0000000000ccd8cf in CacheVC::openWriteStartDone(int, Event*) () at ../../../../../../_vcs/trafficserver9/iocore/cache/CacheWrite.cc:1579
#1  0x0000000000faa6b9 in handleEvent (data=0x609000100c10, event=2, this=<optimized out>) at ../../../../../../_vcs/trafficserver9/iocore/eventsystem/I_Continuation.h:167
#2  handleEvent (data=0x609000100c10, event=2, this=<optimized out>) at ../../../../../../_vcs/trafficserver9/iocore/eventsystem/I_Continuation.h:163
#3  EThread::process_event(Event*, int) () at ../../../../../../_vcs/trafficserver9/iocore/eventsystem/UnixEThread.cc:132
#4  0x0000000000facc9c in EThread::execute_regular (this=this@entry=0x2b28e3c11800) at ../../../../../../_vcs/trafficserver9/iocore/eventsystem/UnixEThread.cc:241
#5  0x0000000000faddee in execute (this=0x2b28e3c11800) at ../../../../../../_vcs/trafficserver9/iocore/eventsystem/UnixEThread.cc:332
#6  EThread::execute (this=0x2b28e3c11800) at ../../../../../../_vcs/trafficserver9/iocore/eventsystem/UnixEThread.cc:310
#7  0x0000000000fa812b in spawn_thread_internal (a=0x6060000a7fc0) at ../../../../../../_vcs/trafficserver9/iocore/eventsystem/Thread.cc:92
#8  0x00002b28d975eea5 in start_thread (arg=0x2b28ea573700) at pthread_create.c:307
#9  0x00002b28da69c8dd in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
(gdb) print od
$1 = (OpenDirEntry *) 0xbebebebebebebebe
```

**Comment by shinrich** [comment]
initializing od to nullptr in the definition of CacheVC (P_CacheInternal.h), seems to have solved the start up crash.

@SolidWallOfCode pointed out that the constructor 0's "part" of the new object.  Is it possible that sizeof(CacheVC) changes with this PR?  I don't see how, but perhaps I'm missing something.  Otherwise, I don't see how this initialization logic differs with and without this patch.

```
CacheVC::CacheVC()
{
  size_to_init = sizeof(CacheVC) - (size_t) & ((CacheVC *)nullptr)->vio;
  memset((void *)&vio, 0, size_to_init);
}
```

File: include/tscore/Allocator.h [hunk 4]
```diff
@@ -108,29 +109,39 @@ class Allocator
...
+  /** Allocates objects of the templated type.  Arguments are forwarded to the constructor for the object. */
+  template <typename... Args>
   C *
-  alloc()
+  alloc(Args &&... args)
   {
     void *ptr = ink_freelist_new(this->fl);
 
-    memcpy(ptr, (void *)&this->proto.typeObject, sizeof(C));
+    ::new (ptr) C(std::forward<Args>(args)...);
```
Comment (ywkaras) [inline include/tscore/Allocator.h:144]: "proto" is a data member of the class allocator, and I think they are always statically allocated.  So the raw memory for proto will be zeroed out at load time before the proto object is constructed by placement new.  We could zero the memory here before calling placement new.  If we don't, there is a risk of seeing further crashes with this change, like the the Susan found.

Comment (shinrich) [reply include/tscore/Allocator.h:144]: Running with ASAN seems to initialize the allocated memory to a highly probably "bad" pattern.  Once I fixed the cacheVC, the rest of the system seems to be running fine.  At least this explains why we didn't see the CacheVC::od problem before.

**Comment by shinrich** [comment]
My test build ran all weekend.  The numbers (rps and average latency) look slightly better than its peer.  Once the change is made in the CacheVC constructor to initialized the od member to nullptr, I'd endorse this change to be merged.

**Comment by ywkaras** [comment]
> My test build ran all weekend. The numbers (rps and average latency) look slightly better than its peer. Once the change is made in the CacheVC constructor to initialize the od member to nullptr, I'd endorse this change to be merged.

Should this change be made as a part of this PR?

**Comment by shinrich** [comment]
Yes, I think this change should be made on this PR.  Otherwise, this PR will break master.

**Comment by ywkaras** [comment]
> Yes, I think this change should be made on this PR. Otherwise, this PR will break master.

Done

**Review by shinrich — DISMISSED** [review]

**Comment by ywkaras** [comment]
Hey @masaori335 can you take another look at this?

File: iocore/eventsystem/I_ProxyAllocator.h [hunk 5]
```diff
@@ -45,84 +48,49 @@ struct ProxyAllocator {
...
-  } else if (count > 0) {
-    a.free_bulk(head, tail, count);
-  }
+#else
 
-  ink_assert(l.allocated >= thread_freelist_low_watermark);
-}
+// Potentially empty varaiable arguments -- Standard way
+//
+#define THREAD_ALLOC(_a, _t, ...) thread_alloc(::_a, _t->_a __VA_OPT__(, ) __VA_ARGS__)
```
Comment (masaori335) [inline iocore/eventsystem/I_ProxyAllocator.h:79]: note: We can't go with this "Standard" way until C++20. (`__VA_OPT__` is introduced by C++20)

**Comment by masaori335** [comment]
@bryancall Any performance differences on your benchmark?

File: iocore/eventsystem/I_ProxyAllocator.h [hunk 6]
```diff
@@ -45,84 +48,49 @@ struct ProxyAllocator {
...
-  if (!cmd_disable_pfreelist && l.freelist) {
-    C *v       = (C *)l.freelist;
-    l.freelist = *(C **)l.freelist;
-    --(l.allocated);
-    memcpy((void *)v, (void *)&a.proto.typeObject, sizeof(C));
-    return v;
-  }
-  return a.alloc();
-}
+void *thread_alloc(Allocator &a, ProxyAllocator &l);
```
Comment (masaori335) [inline iocore/eventsystem/I_ProxyAllocator.h:65]: Bit off-topic, but could you include `tscore/Allocator.h` or add the forward declaration of `Allocator`?

**Comment by masaori335** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
> @bryancall Any performance differences on your benchmark?

Note above that Susan ran this on a production proxy and saw no loss of performance.

**Comment by masaori335** [comment]
> Note above that Susan ran this on a production proxy and saw no loss of performance.

OK, let's ship this.

**Review by masaori335 — APPROVED** [review]

**Comment by shinrich** [comment]
Marking for 9.1 because it makes merging easier.   Especially picking over the h2 to origin branch.

**Comment by zwoop** [comment]
Cherry-picked to v9.1.x branch.

**Pull request merged**